rust/chains/tw_polkadot/src/call_encoder/mod.rs

```diff
@@ -0,0 +1,203 @@
+use crate::ctx_from_tw;
+use tw_proto::Polkadot::Proto::{
+    self,
+    mod_Balance::{BatchAssetTransfer, BatchTransfer, OneOfmessage_oneof as BalanceVariant},
+    mod_CallIndices::OneOfvariant as CallIndicesVariant,
+    mod_SigningInput::OneOfmessage_oneof as SigningVariant,
+    mod_Staking::{
+        Bond, BondAndNominate, Chill, ChillAndUnbond, Nominate,
+        OneOfmessage_oneof as StakingVariant, Unbond,
+    },
+    Balance, CallIndices, Staking,
+};
+use tw_scale::ToScale;
+use tw_ss58_address::NetworkId;
+use tw_substrate::*;
+
+pub mod generic;
+use generic::*;
+
+pub mod polkadot;
+use polkadot::*;
+
+pub mod polymesh;
+use polymesh::*;
+
+pub fn validate_call_index(call_index: &Option<CallIndices>) -> EncodeResult<CallIndex> {
+    let index = match call_index {
+        Some(CallIndices {
+            variant: CallIndicesVariant::custom(c),
+        }) => Some((c.module_index, c.method_index)),
+        _ => None,
+    };
+    CallIndex::from_tw(index)
+}
+
+pub fn required_call_index(call_index: &Option<CallIndices>) -> EncodeResult<CallIndex> {
+    let index = match call_index {
+        Some(CallIndices {
+            variant: CallIndicesVariant::custom(c),
+        }) => Some((c.module_index, c.method_index)),
+        _ => None,
+    };
+    CallIndex::required_from_tw(index)
+}
+
+pub trait TWPolkadotCallEncoder {
+    fn encode_call(&self, msg: &SigningVariant<'_>) -> EncodeResult<Encoded>;
+    fn encode_batch(&self, calls: Vec<Encoded>) -> EncodeResult<Encoded>;
+}
+
+pub struct CallEncoder {
+    encoder: Box<dyn TWPolkadotCallEncoder>,
+}
+
+impl CallEncoder {
+    pub fn from_ctx(ctx: &SubstrateContext) -> EncodeResult<Self> {
+        let encoder = match ctx.network {
+            NetworkId::POLKADOT => PolkadotCallEncoder::new(ctx),
+            NetworkId::KUSAMA => KusamaCallEncoder::new(ctx),
+            NetworkId::POLYMESH => PolymeshCallEncoder::new(ctx),
```

Comment:
While porting chains to Rust, we try to improve overall experience and codebase. Otherwise there is no much need in the migration.
We should handle this on the Entry level. So there should be separate `tw_polkadot` and `tw_kusama` crates, where you can, for example, provide a &dyn TWPolkadotCallEncoder object to a signer, or something like that.
Or it will be even better to provide a dynamic object that returns call indices for the asked extrinsic.

Please also note that we started using Polkadot blockchain implementation for all Substrate chains. For example, you can sign a transaction for Polymesh even if it's not in `registry.json` by using custom `CustomCallIndices`. But for breaking compatibility, we should use default values if the custom indices are not set for Polkadot or Kusama.

I think there is no need to add `tw_polymesh` chain, but rather add it to the `registry.json` file, and we'll use `CustomCallIndices` for every call.

Reply:
I was planning on splitting `tw_polkadot` into per-chain crates later.  For this PR I didn't want to make major structural changes to how the Rust vs C++ Polkadot backend worked.

For a lot of Substrate chains a single generic backend that allows `CustomCallIndices` would be enough, since they might only need Balance/Staking support.  But for chains like Polymesh we have other calls that are needed  to make using Trust Wallet work for our users.

Reply:
Should I split `tw_polkadot` into 3 different chain impls (polkadot, kusama, polymesh) in this PR?  Also we could have one generic Substrate chain crate that requires custom call indices to support chains that just need balance/staking support.

Reply:
I have started work on moving Polymesh support into its own coin (`tw_polymesh` crate).  Here is a draft PR for that: https://github.com/PolymeshAssociation/wallet-core/pull/2

Right now it still uses the `Polkadot` protobufs definition, but we might want to make a new protobuf for Polymesh.